internal/pubsub/samples/gcmsg/main.go

```diff
@@ -0,0 +1,224 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// gcmsg is a sample application that publishes messages from stdin to an
+// existing topic or receives messages from an existing subscription and
+// sends them to stdout. The name gcmsg is short for Go Cloud Messages.
+package main
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"regexp"
+
+	"github.com/google/subcommands"
+	"gocloud.dev/internal/pubsub"
+)
+
+type pubCmd struct{}
+
+func (*pubCmd) Name() string     { return "pub" }
+func (*pubCmd) Synopsis() string { return "Publish a message to a topic" }
+func (*pubCmd) Usage() string {
+	return `pub <topic>:
+  Read messages from stdin, one per line and send them to <topic>.
+`
+}
+
+func (p *pubCmd) SetFlags(f *flag.FlagSet) {
+}
+
+func (p *pubCmd) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
+	if f.NArg() != 1 {
+		f.Usage()
+		return subcommands.ExitUsageError
+	}
+	topicURL := f.Arg(0)
+	if err := p.pub(ctx, topicURL); err != nil {
+		log.Print(err)
+		return subcommands.ExitFailure
+	}
+	return subcommands.ExitSuccess
+}
+
+func (p *pubCmd) pub(ctx context.Context, topicURL string) error {
+	t, cleanup, err := openTopic(ctx, topicURL)
+	if err != nil {
+		return fmt.Errorf("opening topic: %v", err)
+	}
+	defer cleanup()
+
+	// Get lines from stdin and send them as messages to the topic.
+	fmt.Fprintf(os.Stderr, "Enter messages, one per line to be published to \"%s\".\n", topicURL)
+	scanner := bufio.NewScanner(os.Stdin)
+	for scanner.Scan() {
+		line := scanner.Text()
+		m := &pubsub.Message{Body: []byte(line)}
+		if err := t.Send(ctx, m); err != nil {
+			return err
+		}
+	}
+	return scanner.Err()
+}
+
+// openTopic opens a topic on a supported pubsub provider, given a topic URL in a format
+// specific to gcmsg.
+func openTopic(ctx context.Context, topicURL string) (*pubsub.Topic, func(), error) {
+	u, err := url.Parse(topicURL)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parsing URL: %v", err)
+	}
+	rightPart := topicURL[len(u.Scheme+"://"):]
```

Comment:
Changed over to using a regex for this.

Reply:
I looked at the new version. I'm still confused why you're not using `net/url.Parse`. It seems to do some of this work for you, at least the initial split of the scheme.

Reply:
The only work that needs to be done is to split on "://". Using `net/url.Parse` for this means I would then have to extract all the stuff to the right of "://" using different fields of the `URL` object since apparently "all the stuff to the right of ://" doesn't have a name. Using a regex makes it easy and clear to say what I want. Why block on this?

Reply:
Not insisting.

Reply:
You aren't guaranteed that the characters will actually be `"://"`, though. Note that `url.Parse` will accept a URI like `scheme:opaque?query`, in which case this code would misbehave.

URI/URL parsing is extraordinarily subtle. I would recommend using the `net/url` package consistently, especially if this function is eventually going to be made part of the larger Pub/Sub package. Using this string-based method does not handle percent encoding, for example.

At the very least, we should have an issue open for this.

Reply:
Created issue https://github.com/google/go-cloud/issues/1090.